libsignal-service/src/receiver.rs

```diff
@@ -51,4 +59,89 @@ impl<Service: PushService> MessageReceiver<Service> {
             .await?;
         Ok(MessagePipe::from_socket(ws, stream, credentials))
     }
+
+    pub async fn retrieve_contacts(
+        &mut self,
+        contacts: &crate::proto::sync_message::Contacts,
+    ) -> Result<
+        impl Iterator<Item = Result<Contact, ParseContactError>>,
+        ServiceError,
+    > {
+        if let Some(ref blob) = contacts.blob {
+            use futures::io::AsyncReadExt;
+
+            let mut stream = loop {
+                let r = self.service.get_attachment(&blob).await;
+                match r {
+                    Ok(stream) => break stream,
+                    Err(ServiceError::Timeout { .. }) => {
+                        log::warn!("get_attachment timed out, retrying")
+                    }
```

Comment:
I wonder whether we should "fail after five times" or something with a "try again" error? Usually, these time-outs require some backing off, no? What does the NodeJS impl do with this?

Reply:
I just baked in a super basic "retry in a loop" - I didn't go the exponential backoff ways because: 1. it would tie `libsignal-service` to a Futures executor flavor (since you have to sleep in async), 2. our error reporting is not granular enough to handle errors that are worth retrying that aren't timeouts

Reply:
Perfect. Caller should do exp-backoff if needed then. I just imagined running this loop on a disconnected phone that for some reason calls it "time out" (happens on e.g. 4G). You'd just have a near-100%-CPU loop then. This is better :-)